rust/tool/src/install.rs

```diff
@@ -243,6 +317,61 @@ impl Installer {
     }
 }
 
+/// Stores the source and destination of all artifacts needed to install all generations.
+///
+/// This enables a two step installaton process where all artifacts across all generations are
+/// first collected and then installed. This allows deduplication in the collection phase and thus
+/// reduces the number of accesesses and writes to the ESP.
+///
+/// The underlying data structure to store the paths is a HashMap to map source and destination to
+/// each other while automatically ensuring uniqueness.
+struct GenerationArtifacts {
+    /// Temporary directory that stores all temporary files that are created when building the
+    /// GenerationArtifacts.
+    tempdir: TempDir,
+    /// Mapping of signed files from their destinations to their source.
+    signed_files: HashMap<PathBuf, PathBuf>,
+    /// Mapping of unsigned files from their destinations to their source.
+    unsigned_files: HashMap<PathBuf, PathBuf>,
+}
+
+impl GenerationArtifacts {
+    fn new() -> Result<Self> {
+        Ok(Self {
+            tempdir: TempDir::new().context("Failed to create temporary directory.")?,
```

Comment:
A hint that we are abusing this data structure is that you carry a member around that you never touch in the methods below.